controllers/confidential_handler.go

```
@@ -57,22 +77,98 @@ func (r *KataConfigOpenShiftReconciler) handleFeatureConfidential(state FeatureG
+		var peerpodsCMData map[string]string
+		if state == Enabled {
+			peerpodsCMData = map[string]string{"DISABLECVM": "false"}
+		} else {
+			peerpodsCMData = map[string]string{"DISABLECVM": "true"}
+		}
+		if err := updateConfigMap(r.Client, r.Log, peerpodsCMName, OperatorNamespace, peerpodsCMData); err != nil {
+			if k8serrors.IsNotFound(err) {
+				// When feature is Enabled: ConfigMap doesn't exist yet, will try again at the next reconcile run.
+				// Else: Nothing to do, feature is disabled and configMap doesn't exist.
+			} else {
+				return err
+			}
+		}
+	}
 
-	// Patch peer pods configMap, if it exists.
-	var peerpodsCMData map[string]string
+	// Handle kata-cc runtime classes for baremetal environments
 	if state == Enabled {
-		peerpodsCMData = map[string]string{"DISABLECVM": "false"}
+		r.Log.Info("Creating kata-cc runtime classes for confidential containers")
+
+		handler, nodeLabel, err := r.computeTEEHandlerAndLabel()
+		if err != nil {
+			r.Log.Info("TEE detection failed", "err", err)
+			return fmt.Errorf("TEE detection failed: %w", err)
+		}
+
+		// Create kata-cc runtime class restricted to the detected TEE subset
+		err = r.createRuntimeClass(kataCCRuntimeClassName, kataCCRuntimeClassCpuOverhead, kataCCRuntimeClassMemOverhead, handler, nodeLabel)
+		if err != nil {
+			r.Log.Info("Error creating kata-cc runtime class", "err", err)
+			return fmt.Errorf("Error creating kata-cc runtime class: %w", err)
+		}
+
 	} else {
-		peerpodsCMData = map[string]string{"DISABLECVM": "true"}
-	}
-	if err := updateConfigMap(r.Client, r.Log, peerpodsCMName, OperatorNamespace, peerpodsCMData); err != nil {
-		if k8serrors.IsNotFound(err) {
-			// When feature is Enabled: ConfigMap doesn't exist yet, will try again at the next reconcile run.
-			// Else: Nothing to do, feature is disabled and configMap doesn't exist.
-		} else {
-			return err
+		r.Log.Info("Deleting kata-cc runtime classes for confidential containers")
+
+		// Delete kata-cc runtime class
```

I'm still a bit concerned that we can delete the runtime class even though it might still be used. Currently running pods will be unaffected but it will prevent new pods from being created using that runtime class and can cause instability during rollouts or scaling events in deployments and daemonsets... This badly breaks the end user experience IMO.

---

I see the usability issue you mentioned. A few alternatives we could consider:

1. Add a pod validation webhook to reject pods using kata-cc when confidential is disabled. Not sure about the operational implications of adding another webhook though, specially for every new pod. I feel it adds some overhead we don't need.

2. Keep the runtime class but update the handler to "kata". I don't like this either, it silently masks the issue and users might think they're getting confidential computing when they're not.

3. Update the existing KataConfig validation webhook to block disabling confidential if there are pods running with kata-cc or kata-remote. This would force users to migrate workloads first.

Wdyt?

---

As we discussed today, I implemented a finalize approach. Vide last commits and let me know if you spot any issues.